### PR TITLE
Stages refactored to real XBlocks

### PR DESCRIPTION
**Pull request by e-kolpakov**

### CODE REVIEW

File: group_project_v2/stage.py [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+
+        return webob.response.Response(body=json.dumps(results))
+
+    @XBlock.json_handler
+    def submit_peer_feedback(self, submissions, suffix=''):  # pylint: disable=unused-argument
+        try:
+            peer_id = submissions["review_subject_id"]
+            del submissions["review_subject_id"]
+
+            # Then something like this needs to happen
```
Comment (Kelketek) [inline group_project_v2/stage.py]: Something like this needs to happen? Is this not representative of final functionality, then?

File: group_project_v2/stage_components.py [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+            'question_content': self.render_content()
+        }
+        render_context.update(context)
+        fragment.add_content(loader.render_template("templates/html/components/review_question.html", render_context))
+        return fragment
+
+    def studio_view(self, context):
+        fragment = super(GroupProjectReviewQuestionXBlock, self).studio_view(context)
+
+        # TODO: StudioEditableXBlockMixin should really support Codemirror XML editor
```
Comment (Kelketek) [inline group_project_v2/stage_components.py]: Will you be adding that as part of Real Children Part II?

**Comment by Kelketek** [comment]
@e-kolpakov When importing the group work example course you gave me, I get this traceback:

```
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 167, in definition_from_xml
    child_block = system.process_xml(etree.tostring(child, encoding='unicode'))
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml.py", line 177, in process_xml
    id_manager,
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1386, in xblock_from_node
    block = block_class.parse_xml(node, self, keys, id_generator)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/mixins.py", line 386, in parse_xml
    block.runtime.add_node_as_child(block, child, id_generator)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml.py", line 256, in add_node_as_child
    child_block = self.process_xml(etree.tostring(node))
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml.py", line 177, in process_xml
    id_manager,
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1386, in xblock_from_node
    block = block_class.parse_xml(node, self, keys, id_generator)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/mixins.py", line 386, in parse_xml
    block.runtime.add_node_as_child(block, child, id_generator)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml.py", line 256, in add_node_as_child
    child_block = self.process_xml(etree.tostring(node))
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml.py", line 177, in process_xml
    id_manager,
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1389, in xblock_from_node
    block.save()
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/mixins.py", line 248, in save
    fields_to_save = self._get_fields_to_save()
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/mixins.py", line 272, in _get_fields_to_save
    fields_to_save[field.name] = field.to_json(self._field_data_cache[field.name])
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/fields.py", line 865, in to_json
    raise TypeError("Value stored must be a datetime object, not {}".format(type(value)))
TypeError: Value stored must be a datetime object, not <type 'str'>
```

File: group_project_v2/stage_components.py (hunk 2 above)
Comment (e-kolpakov) [inline group_project_v2/stage_components.py]: @Kelketek no, I've hacked together a good-enough solution and I don' think I have time for this in scope of this task.

File: group_project_v2/stage.py (hunk 1 above)
Comment (e-kolpakov) [inline group_project_v2/stage.py]: @Kelketek it's a stale comment from where this method lived previously. I'll remove it.

**Comment by e-kolpakov** [comment]
@Kelketek imported correctly just now. Try installing xblock-utils from https://github.com/edx/xblock-utils/pull/22 - maybe it would fix it. Anyway, let me know when you're available today so we could resolve this issue.

File: group_project_v2/stage.py [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+        help=_(U"This is a name of the stage"),
+        scope=Scope.content,
+        default="Group Project V2 Stage"
+    )
+
+    open_date = DateTime(
+        display_name=_(u"Open Date"),
+        help=_(u"Stage open date"),
+        scope=Scope.settings
+    )
```
Comment (antoviaque) [inline group_project_v2/stage.py]: How do I edit the date field? I've first tried entering "2015-07-10" but I got in the JS console: 

```
SyntaxError: JSON.parse: unexpected non-whitespace character after JSON data at line 1 column 5 of the JSON data"
```

When I tried a JSON time string, "2012-04-23T18:25:43.511Z", it saved, but is now showing this error when I try to edit the properties again:

```
We're having trouble rendering your component
Students will not be able to access this component. Re-edit your component to fix the error.
Error: datetime.datetime(2012, 4, 23, 18, 25, 43, 511000, tzinfo=<UTC>) is not JSON serializable 
```

Comment (e-kolpakov) [inline group_project_v2/stage.py]: @antoviaque it supposed to show a datepicker when focused. Probably installation issue as it is handled by xblock-utils. I'll get back to fresh install as soon as I finish what I'm doing right now.

**Comment by antoviaque** [comment]
When I try to add a Group Assessment in an activity I get the following traceback:

```
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/preview.py", line 281, in get_preview_fragment
    fragment = module.render(preview_view, context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/core.py", line 155, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1770, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1140, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock-utils/xblockutils/studio_editable.py", line 303, in author_view
    return self.author_preview_view(context)
  File "/edx/app/edxapp/edx-platform/dev/xblock-group-project-v2/group_project_v2/stage.py", line 148, in author_preview_view
    return self._view_render(context, "author_view")
  File "/edx/app/edxapp/edx-platform/dev/xblock-group-project-v2/group_project_v2/stage.py", line 130, in _view_render
    stage_fragment = self.get_stage_content_fragment(context, view)
  File "/edx/app/edxapp/edx-platform/dev/xblock-group-project-v2/group_project_v2/stage.py", line 621, in get_stage_content_fragment
    final_grade = self.activity.calculate_grade(self.workgroup['id'])
  File "/edx/app/edxapp/edx-platform/dev/xblock-group-project-v2/group_project_v2/group_project.py", line 356, in calculate_grade
    review_item_data = project_api.get_workgroup_review_items_for_group(group_id, self.content_id)
  File "/edx/app/edxapp/edx-platform/dev/xblock-group-project-v2/group_project_v2/api_error.py", line 59, in call_api_method
    raise api_error
ApiError: ApiError 'INTERNAL SERVER ERROR' (500)
```

Am I missing an installation step?

File: group_project_v2/stage.py (hunk 3 above)
Comment (antoviaque) [inline group_project_v2/stage.py]: @e-kolpakov That was it - thanks!

**Comment by e-kolpakov** [comment]
@antoviaque can't reproduce the issue with adding Group Assessment to an activity. Judging by the traceback, there should be a corresponding traceback in LMS - can you please check this out again and post LMS traceback, if available?

**Comment by antoviaque** [comment]
@e-kolpakov Oh, it needed the LMS - of course. I hadn't checked it, the CMS usually doesn't depend on it. There was some stale `.pyc` files that crashed every request, fixed. I'll test some more, thanks!

**Comment by Kelketek** [comment]
1. When I click on the edit button for the Group Project V2 block on your imported example, the title is huuuuuge. The same is true for the activity blocks, especially the second activity.

2. The CSS also causes the studio to render some of the contents in the lower left of the page, like fixed positioning or something is placing it there. You can scroll down and see it at the very bottom of the page.

Date picking seems happy now, at least.

When I try to load Groupwork in Apros, I get this traceback in the LMS:

```
2015-07-15 01:40:54,990 ERROR 3074 [django.request] base.py:213 - Internal Server Error: /courses/Sandbox/Sandbox/Sandbox/xblock/i4x:;_;_Sandbox;_Sandbox;_group-project-v2-activity;_9792b472400c4f78974156d473298643/view/student_view
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 109, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 1017, in xblock_view
    fragment = instance.render(view_name, context=request.GET)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/core.py", line 155, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1770, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1140, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock-group-project-v2/group_project_v2/group_project.py", line 301, in student_view
    "initialization_data": json.dumps(self._get_initialization_data()),
  File "/edx/app/edxapp/venvs/edxapp/src/xblock-group-project-v2/group_project_v2/group_project.py", line 231, in _get_initialization_data
    "default_stage_id": unicode(self._get_default_stage_id())
  File "/edx/app/edxapp/venvs/edxapp/src/xblock-group-project-v2/group_project_v2/group_project.py", line 235, in _get_default_stage_id
    default_stage_id = self.stages[0].id
IndexError: list index out of range
```

**Comment by e-kolpakov** [comment]
@antoviaque yes, it still needs some work to properly isolate studio views from LMS. There are two ways to do so:

* Make sure only author views are rendered in studio and those views does not query `project_api` in any way.
* Substitute project_api with fake in Studio somehow.

Not sure which approach wuld work best yet.

**Comment by e-kolpakov** [comment]
@Kelketek 

* Title: the title includes all the titles of all the blocks down the road. Looks like a Studio or xblock-utils bug (or maybe I'm just cooking it wrong).
* CSS: thank you - I'll take a look on it.
* Apros: it assumes there's exactly one project navigator in group project. If there are more of them it just ignores all except first, but if there aren't any it crashes. Needs a fix on apros side. I'm aware of it for quite some time, just don't get to fixing it.

**Comment by e-kolpakov** [comment]
@Kelketek update:

* Title: turned out to be a potential bug in studio: basically block editor picks titles of all the XBlocks rendered under it's parent. Those titles are rendered in Studio by PreviewSystem wrapper that applies to both student and author views. So I had to invent "preview_view" to just avoid that wrapping (and fix titles).
* CSS - fixed. It was other group submissions popup
* Apros - fixed, but you'd need to update to latest Apros version in `group_project_v2` branch

**Comment by e-kolpakov** [comment]
@antoviaque I've reworked project_api so that it does not send any requests to LMS when is studio context, so you can now run it without LMS.

**Comment by antoviaque** [comment]
@e-kolpakov FYI, I've tested some more today, and it went well. I mostly focused on the Studio part for now, since that's what I'm going to focus on for their first demo -- I'll test more on the LMS/apros side once the current work from this week is done. There are some details that I might comment on later on (for example, the naming of the peer/group stages is quite confusing at the moment, I was often finding myself scratching my head wondering what was doing what), but I'll see first how it goes with McKinsey during the demo. I didn't see anything blocking so far.

Nice work overall, it's great to see this : ) Kudos!

**Comment by Kelketek** [comment]
:+1:

**Pull request merged**